File: app/src/main/java/com/example/jacob/myapplication/tasks/CreateDataTask.java
package com.example.jacob.myapplication.tasks;

import android.app.Activity;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.ParcelFileDescriptor;
import android.view.View;

import com.example.jacob.myapplication.activities.ResultsActivity;
import com.example.jacob.myapplication.Constants;
import com.example.jacob.myapplication.logic.AnalyzedLine;
import com.example.jacob.myapplication.logic.ConversationData;
import com.example.jacob.myapplication.logic.LineAnalyzer;
import com.example.jacob.myapplication.R;

import java.io.BufferedReader;
import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;

/**
 * Created by jacob on 21/11/2015.
 */
public class CreateDataTask extends AsyncTask<ParcelFileDescriptor, Void, ConversationData> {

    Activity myAct;
    String fileName;

    public CreateDataTask(Activity activity, String name){
        myAct = activity;
        fileName = name;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        myAct.findViewById(R.id.saved_analysis_list).setVisibility(View.INVISIBLE);
        //Debug.startMethodTracing("analysis");
        myAct.findViewById(R.id.loadingPanel).setVisibility(View.VISIBLE);
    }

    @Override
    protected ConversationData doInBackground(ParcelFileDescriptor...  params) {
        return openFile(params[0]);
    }

    @Override
    protected void onPostExecute(ConversationData o) {
        super.onPostExecute(o);
        if (o != null){
            myAct.findViewById(R.id.saved_analysis_list).setVisibility(View.VISIBLE);
            myAct.findViewById(R.id.loadingPanel).setVisibility(View.GONE);
            Constants.conversationData = o;
            Intent intent = new Intent(myAct, ResultsActivity.class);
            myAct.startActivity(intent);
        }
    }

    private ConversationData openFile(ParcelFileDescriptor arg) {
        FileDescriptor fd = arg.getFileDescriptor();
        BufferedReader bf = null;
        try{
            bf = new BufferedReader (new InputStreamReader(new FileInputStream(fd), "utf-8"));
        }
        catch (Exception e){
            System.err.println(e);
        }
        LineAnalyzer analyzer = new LineAnalyzer();
        ConversationData data = new ConversationData(fileName);
        String line;
        try{
            while ((line = bf.readLine()) != null) {
                AnalyzedLine a = analyzer.analyzeLine(line);
                if (a!= null)
                    data.addData(a);
            }
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
        data.createTotalDaysData();
        data.createMonthsData();
        return data;
    }
}

File: app/src/main/java/me/jacobrr/whatsappanalyzer/tasks/CreateDataTask.kt
package me.jacobrr.whatsappanalyzer.tasks

import android.app.Activity
import android.content.Intent
import android.os.AsyncTask
import android.os.ParcelFileDescriptor
import android.view.View
import me.jacobrr.whatsappanalyzer.Constants
import me.jacobrr.whatsappanalyzer.R
import me.jacobrr.whatsappanalyzer.activities.ResultsActivity
import me.jacobrr.whatsappanalyzer.logic.ConversationData
import me.jacobrr.whatsappanalyzer.logic.LineAnalyzer

import java.io.*

/**
 * Created by jacob on 21/11/2015.
 */
class CreateDataTask(internal var myAct: Activity, internal var fileName: String) : AsyncTask<ParcelFileDescriptor, Void, ConversationData>() {

    override fun onPreExecute() {
        super.onPreExecute()
        myAct.findViewById<View>(R.id.saved_analysis_list).visibility = View.INVISIBLE
        //Debug.startMethodTracing("analysis");
        myAct.findViewById<View>(R.id.loadingPanel).visibility = View.VISIBLE
    }

    override fun doInBackground(vararg params: ParcelFileDescriptor): ConversationData {
        return openFile(params[0])
    }

    override fun onPostExecute(o: ConversationData?) {
        super.onPostExecute(o)
        if (o != null) {
            myAct.findViewById<View>(R.id.saved_analysis_list).visibility = View.VISIBLE
            myAct.findViewById<View>(R.id.loadingPanel).visibility = View.GONE
            Constants.conversationData = o
            val intent = Intent(myAct, ResultsActivity::class.java)
            myAct.startActivity(intent)
        }
    }

    private fun openFile(arg: ParcelFileDescriptor): ConversationData {
        val fd = arg.fileDescriptor
        var bf: BufferedReader? = null
        try {
            bf = BufferedReader(InputStreamReader(FileInputStream(fd), "utf-8"))
        } catch (e: Exception) {
            System.err.println(e)
        }

        val analyzer = LineAnalyzer()
        val data = ConversationData(fileName)

        try {
            var line: String? = bf!!.readLine()
            while (line != null) {
                val a = analyzer.analyzeLine(line)
                if (a != null)
                    data.addData(a)
                line = bf.readLine()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        data.createTotalDaysData()
        data.createMonthsData()
        return data
    }
}